experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,140 @@
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut requests_to_shuffle = self
+                .requests_to_shuffle
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+
+            let request = Message {
+                number: requests_to_shuffle.len(),
+                data: request,
+                response_sender,
+            };
+
+            requests_to_shuffle.push(request);
+
+            if requests_to_shuffle.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(requests_to_shuffle.deref_mut(), vec![]);
+
+                // Shuffle requests and spawn async tasks for sending them to the backend.
+                let request_handler = self.request_handler.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_handler).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        response_receiver.await.context("Couldn't receive response")
+    }
+
+    // Lexicographically sorts requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(
+        mut requests: Vec<Message>,
+        request_handler: Arc<dyn RequestHandler>,
+    ) -> anyhow::Result<()> {
+        requests.sort_by_key(|v| v.data.clone());
+
+        let response_futures = requests.into_iter().map(|request| {
+            let request_handler_clone = request_handler.clone();
+            async move {
+                let response = request_handler_clone.handle(request.data).await;
+                Message {
+                    number: request.number,
+                    data: response,
+                    response_sender: request.response_sender,
+                }
+            }
+        });
+        let responses = join_all(response_futures).await;
+
+        Self::shuffle_responses(responses)
+    }
+
+    // Restores the original order in which messages arrived and sends them back to the client async
+    // tasks using corresponding channels defined in the [`Message::response_sender`].
+    fn shuffle_responses(mut responses: Vec<Message>) -> anyhow::Result<()> {
+        responses.sort_by(|first, second| first.number.cmp(&second.number));
+
+        responses
+            .into_iter()
+            .map(
+                |Message {
+                     number: _,
+                     data,
+                     response_sender,
+                 }| {
+                    response_sender
+                        .send(data)
+                        .map_err(|_| anyhow!("Couldn't send response"))
```

Comment:
Why do we just throw away the error here? Doesn't that make it hard to debug?

Reply:
Thanks, perhaps something like this may clarify

```suggestion
                        .map_err(|_data| anyhow!("Couldn't send response"))
```